doc/man/pam_end.3
.\" Generated from pam_end.c by gendoc.pl
.\" $Id: pam_end.c 913 2017-01-21 15:11:12Z des $
.Dd February 19, 2017
.Dt PAM_END 3
.Os
.Sh NAME
.Nm pam_end
.Nd terminate the PAM transaction
.Sh SYNOPSIS
.In sys/types.h
.In security/pam_appl.h
.Ft "int"
.Fn pam_end "pam_handle_t *pamh" "int status"
.Sh DESCRIPTION
The
.Fn pam_end
function terminates a PAM transaction and destroys the
corresponding PAM context, releasing all resources allocated to it.
.Pp
The
.Fa status
argument should be set to the error code returned by the
last API call before the call to
.Fn pam_end .
.Sh RETURN VALUES
The
.Fn pam_end
function returns 0 on success and -1 on failure.
.Sh SEE ALSO
.Xr pam 3 ,
.Xr pam_strerror 3
.Sh STANDARDS
.Rs
.%T "X/Open Single Sign-On Service (XSSO) - Pluggable Authentication Modules"
.%D "June 1997"
.Re
.Sh AUTHORS
The
.Fn pam_end
function and this manual page were
developed for the
.Fx
Project by ThinkSec AS and Network Associates Laboratories, the
Security Research Division of Network Associates, Inc.\& under
DARPA/SPAWAR contract N66001-01-C-8035
.Pq Dq CBOSS ,
as part of the DARPA CHATS research program.
.Pp
The OpenPAM library is maintained by
.An Dag-Erling Sm\(/orgrav Aq Mt des@des.no .


NEWS
$Id: NEWS,v 1.20 2017/02/16 14:38:12 schwarze Exp $

This file lists the most important changes in the mdocml.bsd.lv distribution.

Changes in version 1.14.1, released on February 21, 2017

    --- MAJOR NEW FEATURES ---
 * apropos(1): Reimplement complete semantic search functionality
   without the dependency on SQLite3, using only POSIX APIs.
   This comes with a completely new mandoc.db(5) file format.
 * man(1): Support more than one tag entry for the same search term,
   plus some minor improvements to the less(1) :t support.
 * -Thtml: Use real macro names for CSS classes.
   Systematic cleanup of and many improvements to mandoc.css.
 * -Thtml: Produce human readable HTML code by using indentation
   and better line breaks.  Improve various HTML elements,
   and trim several useless ones.
 * New catman(8) utility, still somewhat experimental.
 * Now includes a portable version of the OpenBSD mandoc regression
   suite, see regress/regress.pl.1 for details.
    --- REMOVED FUNCTIONALITY ---
 * Operating systems that don't provide mmap(3) are no longer supported.
 * Drop support for manpath(1).  Even if your system has manpath(1),
   it is simpler to use MANPATH_DEFAULT in configure.local for
   operating system defaults, man.conf(5) for machine-specific
   modifications, and ${MANPATH}, -m, and -M for user preferences
   than to bother with the complexity of manpath(1).
 * makewhatis(8) -p: No longer warn about missing MLINKS since these
   are no longer needed for anything.
    --- MINOR NEW FEATURES ---
 * mdoc(7): Warn about invalid punctuation and content below NAME.
 * mdoc(7): Warn about .Xr lacking the second argument (section).
 * mdoc(7): Warn about violations of the rule "new sentence, new line".
 * roff(7): Warn about trailing whitespace at the end of comments.
 * mdoc(7): Improve rendering of double quotes.
 * mdoc(7): Always do text production in the validator, never in the
   formatters.  Cleaner, simpler, shorter, helps NetBSD apropos(1)
   and also makes -Ttree output more useful.
 * -Ttree: Show metadata and some additional node flags.
   New -Onoval output option to show the unvalidated tree.
    --- RELIABILITY BUGFIXES ---
 * man(1): Make "man -l" work with standard input from a pipe or file,
   as long as standard output is a terminal.
 * man(7): Fix out of bounds read access if a text node immediately
   preceded the first .SH header.
 * mdoc(7): Fix out of bounds read access for .Bl without a type
   but with a width.
 * mdoc(7): Fix out of bounds read access for .Bl -column starting
   with a tab character instead of a child .It macro.
 * mdoc(7): Fix syntax tree corruption leading to segfaults caused
   by stray block end macros in nested blocks of mismatching type.
 * man(1): Fix NULL dereference when the first of multiple pages
   shown was preformatted.
 * mdoc(7): Fix syntax tree corruption leading to NULL dereference
   caused by partial implicit macros inside .Bl -column table cells.
 * mdoc(7): Fix syntax tree corruption leading to NULL dereference
   for macro sequences like .Bl .Bl .It Bo .El .It.
 * mdoc(7): Fix syntax tree corruption leading to NULL dereference
   caused by .Ta following a nested .Bl -column breaking another block.
 * mdoc(7): Fix syntax tree corruption sometimes leading to NULL
   dereference caused by indirectly broken .Nd or .Nm blocks.
 * mdoc(7) -Thtml: Fix a NULL dereference for .Bl -column with 0 columns.
 * mdoc(7): Fix NULL dereference in some specific cases of a
   block-end macro calling another block-end macro.
 * mdoc(7): Fix NULL dereference if the only child of the head
   of the first .Sh was an empty in-line macro.
 * eqn(7): Fix NULL dereference in the terminal formatter
   for empty matrices and empty square roots.
 * mdoc(7): Fix an assertion failure for a .Bd without a type that
   breaks another block.
 * mdoc(7): Fix an assertion failure that happened for some .Bl -column
   lists containing a column width of "-4n", "-3n", or "-2n".
 * mdoc(7): Fix an assertion failure caused by .Bl -column without .It
   but containing eqn(7) or tbl(7) code.
 * roff(7): Fix an assertion failure caused by \z\[u00FF] with -Tps/-Tpdf.
 * roff(7): Fix an assertion failures caused by whitespace inside \o''
   (overstrike) sequences.
 * -Thtml: Fix an assertion failure caused by -Oman or -Oincludes of
   excessive length.
    --- PORTABILITY IMPROVEMENTS ---
 * man(1): Do not mix stdio narrow and wide stream orientation
   on stdout, which could cause output corruption on glibc.
 * mandoc(1): Autodetect a suitable locale for -Tutf8 mode.
 * ./configure: Autodetect whether PATH_MAX and O_DIRECTORY are defined.
 * ./configure: Autodetect if nanosleep(3) needs -lrt.
 * ./configure: Provide an ${LN} configuration variable.
 * ./configure: Put compiler arguments that may contain -l at the end.
    --- MINOR BUGFIXES ---
 * mdoc(7): Fix SYNOPSIS output if the first child of .Nm is a macro.
 * mdoc(7) -Thtml: Improve formatting of .Bl -tag with short tags.
 * man(7) -Thtml: Preserve whitespace in .nf (nofill) mode.
 * mandoc(1): Error out on invalid output options on the command line.
    --- STRUCTURAL CHANGES, no functional change ---
 * Redesign part of the mandoc_html(3) interfaces, making them much
   easier to use and reducing the amount of code by a few hundred lines.
    --- THANKS TO ---
 * Michael Stapelberg (Debian) for designing the new mandocd(8)
   and parts of the new catman(8), for release testing, and for a
   number of patches and bug reports.
 * Baptiste Daroussin (FreeBSD) for profiling the new makewhatis(8)
   implementation and suggesting an algorithmic improvement which
   more than doubled performance, and for a few bug reports.
 * Ed Maste (FreeBSD) for an important patch improving reproducibility
   of builds in makewhatis(8), and for a few bug reports.
 * Theo Buehler (OpenBSD) for almost twenty important bug reports,
   most of them found by systematic afl(1) fuzzing.

lib/libpam/pam_end.c
/*-
 * Copyright (c) 2002-2003 Networks Associates Technology, Inc.
 * Copyright (c) 2004-2011 Dag-Erling Smørgrav
 * All rights reserved.
 *
 * This software was developed for the FreeBSD Project by ThinkSec AS and
 * Network Associates Laboratories, the Security Research Division of
 * Network Associates, Inc.  under DARPA/SPAWAR contract N66001-01-C-8035
 * ("CBOSS"), as part of the DARPA CHATS research program.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id: pam_end.c 913 2017-01-21 15:11:12Z des $
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdlib.h>

#include <security/pam_appl.h>

#include "openpam_impl.h"

/*
 * XSSO 4.2.1
 * XSSO 6 page 42
 *
 * Terminate the PAM transaction
 */

int
pam_end(pam_handle_t *pamh,
	int status)
{
	pam_data_t *dp;
	int i;

	ENTER();

	/* clear module data */
	while ((dp = pamh->module_data) != NULL) {
		if (dp->cleanup)
			(dp->cleanup)(pamh, dp->data, status);
		pamh->module_data = dp->next;
		FREE(dp->name);
		FREE(dp);
	}

	/* clear environment */
	while (pamh->env_count) {
		--pamh->env_count;
		FREE(pamh->env[pamh->env_count]);
	}
	FREE(pamh->env);

	/* clear chains */
	openpam_clear_chains(pamh->chains);

	/* clear items */
	for (i = 0; i < PAM_NUM_ITEMS; ++i)
		pam_set_item(pamh, i, NULL);

	FREE(pamh);

	RETURNC(PAM_SUCCESS);
}

/**
 * The =pam_end function terminates a PAM transaction and destroys the
 * corresponding PAM context, releasing all resources allocated to it.
 *
 * The =status argument should be set to the error code returned by the
 * last API call before the call to =pam_end.
 */
